<a href="https://colab.research.google.com/github/Data-Intelligence-Mastery/data_science_interview_questions/blob/master/Q028_markov_chain_insurance.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Auto insurance customers

`Statistics Transitional Diagram Probability Theory`

An insurance company classifies its customers into 3 categories: below average, average, and above average. No one moves more than one state at a time.

For example, a customer cannot move from below average to above average or from above average to below average in a given period. After a given period, we notice that:

40% of those in the below average category become average

30% of those in the average category become above average

10% of those in the average category become below average

20% of those in the above average category are downgraded to average

Given the above:

Can you write the transition matrix for the model?

What is the limiting fraction of drivers in each of these categories? (e.g. find the steady state)

Hint: It might be helpful to start with building a state transition diagram to depict a Markov chain for the different states a customer can move to given their current state.



In [47]:
# Transition matrix T is defined as below according to the problem description

import numpy as np

T = np.array([[.8, .2, 0],
              [.3, .6, .1],
              [0, .4, .6]])


T

array([[0.8, 0.2, 0. ],
       [0.3, 0.6, 0.1],
       [0. , 0.4, 0.6]])

array([0.54545455, 0.36363636, 0.09090909])

In [54]:
# Method 1
# writing function to find the stead state vector
def steady_state_prop(p):
    dim = p.shape[0]
    q = (p-np.eye(dim))
    ones = np.ones(dim)
    q = np.c_[q,ones]
    QTQ = np.dot(q, q.T)
    bQT = np.ones(dim)
    return np.linalg.solve(QTQ,bQT)
steady_state_prop(T.T)

array([0.33333333, 0.33333333, 0.33333333])

In [52]:
# Method 2
# using numpy.linalg.eig function
np.linalg.eig(T)
# it output eigenvalue w and eigenvector v

(array([1.        , 0.67320508, 0.32679492]),
 array([[-0.57735027,  0.27315957,  0.23188144],
        [-0.57735027, -0.17317623, -0.54863738],
        [-0.57735027, -0.94625252,  0.80326087]]))

The eigenvector v[:,0] corresponds to eigenvalue w[0]=1, rescale the v[:, 0] so its sum is equal to 1, we will get the same answer from method 1:[0.33, 0.33, 0.33]